File: cypress/cypress/pages/CheckoutPage.ts
import { BasePage } from './BasePage';

const SELECTORS = {
  firstName:           '[data-test="firstName"]',
  lastName:            '[data-test="lastName"]',
  postalCode:          '[data-test="postalCode"]',
  continueButton:      '[data-test="continue"]',
  finishButton:        '[data-test="finish"]',
  cancelButton:        '[data-test="cancel"]',
  confirmationHeader:  '.complete-header',
  confirmationText:    '.complete-text',
  errorMessage:        '[data-test="error"]',
} as const;

export class CheckoutPage extends BasePage {
  openStepOne(): void {
    this.visit('/checkout-step-one.html');
  }

  fillInfo(firstName: string, lastName: string, postalCode: string): void {
    // cy.type() throws on empty strings — skip fields intentionally left blank.
    if (firstName) this.type(SELECTORS.firstName, firstName);
    if (lastName) this.type(SELECTORS.lastName, lastName);
    if (postalCode) this.type(SELECTORS.postalCode, postalCode);
  }

  continueToStepTwo(): void {
    this.click(SELECTORS.continueButton);
  }

  finish(): void {
    this.click(SELECTORS.finishButton);
  }

  cancel(): void {
    this.click(SELECTORS.cancelButton);
  }

  getConfirmationMessage(): Cypress.Chainable<string> {
    return this.getText(SELECTORS.confirmationHeader);
  }

  getConfirmationText(): Cypress.Chainable<string> {
    return this.getText(SELECTORS.confirmationText);
  }

  getErrorMessage(): Cypress.Chainable<string> {
    return this.getText(SELECTORS.errorMessage);
  }
}

File: stagehand-agent/pages/checkout_page.py
"""
checkout_page.py — Page Object for the SauceDemo checkout flow.

Covers checkout step one (info form), step two (overview), and
the completion page. Selectors match cypress/cypress/pages/CheckoutPage.ts.
"""

from __future__ import annotations

from pages.base_page import BasePage


class CheckoutPage(BasePage):
    """Checkout flow interactions for SauceDemo."""

    # Step One — Information
    FIRST_NAME = "[data-test='firstName']"
    LAST_NAME = "[data-test='lastName']"
    POSTAL_CODE = "[data-test='postalCode']"
    CONTINUE_BUTTON = "[data-test='continue']"
    ERROR_MESSAGE = "[data-test='error']"

    # Step Two — Overview
    FINISH_BUTTON = "[data-test='finish']"
    SUMMARY_TOTAL = ".summary_total_label"

    # Complete
    COMPLETE_HEADER = ".complete-header"

    def fill_info(self, first_name: str, last_name: str, postal_code: str) -> None:
        """Fill in the checkout information form."""
        self.type_text(self.FIRST_NAME, first_name)
        self.type_text(self.LAST_NAME, last_name)
        self.type_text(self.POSTAL_CODE, postal_code)

    def continue_to_overview(self) -> None:
        """Click 'Continue' to proceed to the order overview."""
        self.click(self.CONTINUE_BUTTON)

    def finish_checkout(self) -> None:
        """Click 'Finish' to complete the order."""
        self.click(self.FINISH_BUTTON)

    def get_confirmation_header(self) -> str:
        """Return the text of the order completion header."""
        return self.get_text(self.COMPLETE_HEADER)

    def get_summary_total(self) -> str:
        """Return the total price string from the overview page."""
        return self.get_text(self.SUMMARY_TOTAL)

    def is_error_displayed(self) -> bool:
        """Return True if a checkout error message is visible."""
        return self.is_element_visible(self.ERROR_MESSAGE)
